### PR TITLE
Dispatcher

### PR DESCRIPTION
**Pull request by dain**
Move queued phase of query from QueryManager to a new dispatcher service. This
change is in preparation for adding a optional new server that moves the queue
phase to a separate process.

Ref prestodb/presto#12176

### CODE REVIEW

**Comment by sopel39** [comment]
> The min worker count is configured with query-manager.required-workers-max-wait

So the semantics is that queries will be queued until there is min number of workers. What if workers leave the cluster after the cluster is started? Will the new queries be on hold too? This is useful in case of downscaling cluster to 0 nodes.

**Comment by dain** [comment]
> > The min worker count is configured with query-manager.required-workers-max-wait
> 
> So the semantics is that queries will be queued until there is min number of workers. What if workers leave the cluster after the cluster is started? Will the new queries be on hold too? This is useful in case of downscaling cluster to 0 nodes for idle periods in cloud environment.

Yes.  The new behavior is designed for cloud environments that scale down to zero when there is no traffic and then scale back up when there are queries.

File: presto-main/src/main/java/io/prestosql/execution/QueryState.java [hunk 1]
```diff
@@ -28,6 +28,10 @@
      * Query is waiting for the required resources (beta).
      */
     WAITING_FOR_RESOURCES(false),
+    /**
```
Comment (raghavsethi) [inline presto-main/src/main/java/io/prestosql/execution/QueryState.java]: Commit message:

`Add DISPATCHING query state`

Would also recommend some more detail about what this state means in the full message.

**Comment by electrum** [comment]
First three commits look good

**Comment by raghavsethi** [comment]
Can we add a commit prior that adds a specific error code (not generic insufficient resources), so that clients can deal with that situation properly?

File: presto-main/src/main/java/io/prestosql/execution/QueryStateMachine.java [hunk 2]
```diff
@@ -98,7 +98,7 @@
     private final String query;
     private final Session session;
     private final URI self;
-    private final Optional<ResourceGroupId> resourceGroup;
```
Comment (raghavsethi) [inline presto-main/src/main/java/io/prestosql/execution/QueryStateMachine.java]: Remote Optional -> Remove Optional

File: presto-main/src/main/java/io/prestosql/dispatcher/DispatchManager.java [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+        ScheduledExecutorService scheduledExecutorService = newScheduledThreadPool(queryManagerConfig.getQueryManagerExecutorPoolSize(), threadsNamed("query-dispatch-%s"));
+        queryManagementExecutor = listeningDecorator(scheduledExecutorService);
+        queryManagementExecutorMBean = new ThreadPoolExecutorMBean((ThreadPoolExecutor) scheduledExecutorService);
+
+        this.queryTracker = new QueryTracker<>(queryManagerConfig, queryManagementExecutor);
+    }
+
+    @Managed(description = "Query dispatch executor")
+    @Nested
+    public ThreadPoolExecutorMBean getExecutor()
```
Comment (raghavsethi) [inline presto-main/src/main/java/io/prestosql/dispatcher/DispatchManager.java]: This export has no effect

File: presto-main/src/main/java/io/prestosql/dispatcher/QueuedStatementResource.java [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+                        .build());
+    }
+
+    private static final class Query
+    {
+        private final String query;
+        private final SessionContext sessionContext;
+        private final DispatchManager dispatchManager;
+        private final QueryId queryId;
+        private final String slug = format("%016x%016x", ThreadLocalRandom.current().nextLong(), ThreadLocalRandom.current().nextLong());
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/dispatcher/QueuedStatementResource.java]: https://github.com/prestosql/presto/pull/561#discussion_r270119287

File: presto-main/src/main/java/io/prestosql/execution/QueryExecution.java [hunk 5]
```diff
@@ -34,22 +35,36 @@
...
+
     void addOutputInfoListener(Consumer<QueryOutputInfo> listener);
 
     Plan getQueryPlan();
 
+    BasicQueryInfo getBasicQueryInfo();
+
     QueryInfo getQueryInfo();
 
+    String getSlug();
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/execution/QueryExecution.java]: https://github.com/prestosql/presto/pull/561#discussion_r270148132

File: presto-main/src/main/java/io/prestosql/dispatcher/QueuedStatementResource.java [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+            this.dispatchManager = requireNonNull(dispatchManager, "dispatchManager is null");
+            this.queryId = dispatchManager.createQueryId();
+        }
+
+        public QueryId getQueryId()
+        {
+            return queryId;
+        }
+
+        public String getSlug()
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/dispatcher/QueuedStatementResource.java]: https://github.com/prestosql/presto/pull/561#discussion_r270148132 ?

File: presto-main/src/main/java/io/prestosql/dispatcher/LocalDispatchQueryFactory.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        this.transactionManager = requireNonNull(transactionManager, "transactionManager is null");
+        this.accessControl = requireNonNull(accessControl, "accessControl is null");
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.queryMonitor = requireNonNull(queryMonitor, "queryMonitor is null");
+        this.locationFactory = requireNonNull(locationFactory, "locationFactory is null");
+        this.executionFactories = requireNonNull(executionFactories, "executionFactories is null");
+        this.warningCollectorFactory = requireNonNull(warningCollectorFactory, "warningCollectorFactory is null");
+
+        Node currentNode = requireNonNull(internalNodeManager, "internalNodeManager is null").getCurrentNode();
+        this.coordinatorLocation = new CoordinatorLocation(Optional.of(currentNode.getHttpUri()), Optional.of(currentNode.getHttpUri()));
```
Comment (raghavsethi) [inline presto-main/src/main/java/io/prestosql/dispatcher/LocalDispatchQueryFactory.java]: For my own reference: here's the bug from yesterday.

File: presto-main/src/main/java/io/prestosql/dispatcher/LocalDispatchQuery.java [hunk 8]
```diff
@@ -131,15 +134,15 @@ public DateTime getLastHeartbeat()
...
-            failureInfo = stateMachine.getFailureInfo();
+            ExecutionFailureInfo failureInfo = stateMachine.getFailureInfo()
+                    .orElseGet(() -> toFailure(new PrestoException(GENERIC_INTERNAL_ERROR, "Query failed for an unknown reason")));
+            return DispatchInfo.failed(failureInfo, queryInfo.getQueryStats().getElapsedTime(), queryInfo.getQueryStats().getQueuedTime());
+        }
+        if (queryInfo.getState().ordinal() >= PLANNING.ordinal()) {
+            return DispatchInfo.dispatched(coordinatorLocation, queryInfo.getQueryStats().getElapsedTime(), queryInfo.getQueryStats().getQueuedTime());
         }
-        return new DispatchInfo(coordinator, failureInfo, queryInfo.getQueryStats().getElapsedTime(), queryInfo.getQueryStats().getQueuedTime());
+        return DispatchInfo.queued(queryInfo.getQueryStats().getElapsedTime(), queryInfo.getQueryStats().getQueuedTime());
```
Comment (raghavsethi) [inline presto-main/src/main/java/io/prestosql/dispatcher/LocalDispatchQuery.java]: Nit? Edge case for static import?

File: presto-main/src/main/java/io/prestosql/dispatcher/DispatchManager.java [hunk 9]
```diff
@@ -135,30 +134,31 @@ public QueryId createQueryId()
...
+            finally {
+                queryCreationFuture.set(null);
             }
         });
         return queryCreationFuture;
     }
 
+    /**
+     *  Creates and registers a dispatch query with the query tracker.  This method will never fail to register a query with the query
+     *  tracker.  If an error occurs while, creating a dispatch query a failed dispatch will be created and registered.
```
Comment (raghavsethi) [inline presto-main/src/main/java/io/prestosql/dispatcher/DispatchManager.java]: Remove comma

File: presto-main/src/main/java/io/prestosql/dispatcher/DispatchExecutor.java [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    public class DispatchExecutorMBeans
+    {
+        private final ThreadPoolExecutorMBean executor;
+        private final ThreadPoolExecutorMBean scheduledExecutor;
+
+        public DispatchExecutorMBeans(ExecutorService coreExecutor, ScheduledExecutorService coreScheduledExecutor)
+        {
+            executor = new ThreadPoolExecutorMBean((ThreadPoolExecutor) coreExecutor);
```
Comment (raghavsethi) [inline presto-main/src/main/java/io/prestosql/dispatcher/DispatchExecutor.java]: requireNonNull

File: presto-main/src/main/java/io/prestosql/dispatcher/DispatchInfo.java [hunk 11]
```diff
@@ -27,7 +27,24 @@
     private final Duration elapsedTime;
     private final Duration queuedTime;
 
-    public DispatchInfo(Optional<CoordinatorLocation> coordinatorLocation, Optional<ExecutionFailureInfo> failureInfo, Duration elapsedTime, Duration queuedTime)
+    public static DispatchInfo queued(Duration elapsedTime, Duration queuedTime)
```
[the remark this answers is not in the document]
Comment (dain) [reply presto-main/src/main/java/io/prestosql/dispatcher/DispatchInfo.java]: I go back and forth on this.  In this case I like the FQN.

File: presto-main/src/main/java/io/prestosql/dispatcher/DispatchExecutor.java [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+import java.util.concurrent.ThreadPoolExecutor;
+
+import static com.google.common.util.concurrent.MoreExecutors.listeningDecorator;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static java.util.concurrent.Executors.newCachedThreadPool;
+import static java.util.concurrent.Executors.newScheduledThreadPool;
+
+public class DispatchExecutor
+{
+    private final Closer closer = Closer.create();
```
[the remark this answers is not in the document]
Comment (dain) [reply presto-main/src/main/java/io/prestosql/dispatcher/DispatchExecutor.java]: I'm not sure what you mean.  This class uses a closer and an `@PreDestroy`

File: presto-main/src/main/java/io/prestosql/dispatcher/LocalDispatchQuery.java [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+import static io.airlift.concurrent.MoreFutures.addExceptionCallback;
+import static io.airlift.concurrent.MoreFutures.addSuccessCallback;
+import static io.airlift.concurrent.MoreFutures.tryGetFutureValue;
+import static io.airlift.units.DataSize.Unit.BYTE;
+import static io.prestosql.execution.QueryState.FAILED;
+import static io.prestosql.execution.QueryState.PLANNING;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class LocalDispatchQuery
```
Comment (wenleix) [inline presto-main/src/main/java/io/prestosql/dispatcher/LocalDispatchQuery.java]: Curious: Why this is called `LocalDispatchQuery` ?

File: presto-main/src/main/java/io/prestosql/dispatcher/DispatchManager.java [hunk 14]
```diff
@@ -0,0 +1,319 @@
...
+        this.queryPreparer = requireNonNull(queryPreparer, "queryPreparer is null");
+        this.resourceGroupManager = requireNonNull(resourceGroupManager, "resourceGroupManager is null");
+        this.dispatchQueryFactory = requireNonNull(dispatchQueryFactory, "dispatchQueryFactory is null");
+        this.failedDispatchQueryFactory = requireNonNull(failedDispatchQueryFactory, "failedDispatchQueryFactory is null");
+        this.transactionManager = requireNonNull(transactionManager, "transactionManager is null");
+        this.accessControl = requireNonNull(accessControl, "accessControl is null");
+        this.sessionSupplier = requireNonNull(sessionSupplier, "sessionSupplier is null");
+        this.sessionPropertyDefaults = requireNonNull(sessionPropertyDefaults, "sessionPropertyDefaults is null");
+
+        this.maxQueryLength = queryManagerConfig.getMaxQueryLength();
```
[the remark this answers is not in the document]
Comment (raghavsethi) [reply presto-main/src/main/java/io/prestosql/dispatcher/DispatchManager.java]: ^

File: presto-main/src/main/java/io/prestosql/dispatcher/DispatchManager.java (hunk 9 above)
Comment (raghavsethi) [reply presto-main/src/main/java/io/prestosql/dispatcher/DispatchManager.java]: ^

**Pull request merged**